### DISCUSSION TITLE
How can i set custom dataset for the different parameters of my api

### DISCUSSION THREAD

**Discussion by gouseferoz** [schemathesis/schemathesis#1332] [Q&A]
Hi,

I have a use case where in my API there are parameters like UUID, Date and these values are set as strings in the API. So while using schemathesis I see Date/UUID is not getting sent instead random strings are sent which is a valid. 

But how can i send custom data for each API. Because few APIs will have UUID but few not (example).

Is there a way to set custom data for the parameters of the API?

I appreciate your help 👍🏻

**Comment by Stranger6667** [comment]
Hi @gouseferoz 

Thank you for asking :)

> I have a use case where in my API there are parameters like UUID, Date and these values are set as strings in the API. So while using schemathesis I see Date/UUID is not getting sent instead random strings are sent which is a valid.

You can utilize the "format" keyword with [custom string strategies](https://schemathesis.readthedocs.io/en/stable/extending.html#custom-string-strategies). For example:

```python
# hooks.py
from hypothesis import strategies

from schemathesis import register_string_format

register_string_format("uuid", strategies.uuids().map(str))
```

For CLI it will require using the `pre-run` hook:

```
schemathesis --pre-run importable.path.to.hooks.py run ...
```

Similarly, it works for date formats, as the [standard string formats](https://swagger.io/specification/#data-types) are supported out of the box.

Hope that it helps :) Let me know if I can elaborate or if adding `format: uuid` to the API schema is not an option.

**Reply by gouseferoz** [reply]
Hi @stranger6667

Thanks for the reply, I was able to use the string format and I see the data being passed for the UUID parameter is something like this: Æ\U000f5724Ø<Î\U000e68bfHá

I am unable to figure out what the above string format is. Can you help me with this?

**Reply by Stranger6667** [reply]
Could you, please, share your API schema? It will help me a lot (it could be anonymized, the "schema" part of parameters & `requestBody` is the most important)

**Reply by gouseferoz** [reply]
I am using the flask-restplus for my api's and below is the schema generated by it.
`{
swagger: "2.0",
basePath: "/",
paths: {
/v1/get/{ident}/{ticker}: {
parameters: [
{
name: "ident",
in: "path",
required: true,
type: "string"
},
{
name: "ticker",
in: "path",
required: true,
type: "integer"
}
],
get: {
responses: {
200: {
description: "Success"
}
},
operationId: "get_hello_world",
tags: [
"v1"
]
}
},
/v1/post: {
post: {
responses: {
200: {
description: "Success"
}
},
operationId: "post_hello_world",
parameters: [
{
name: "payload",
required: true,
in: "body",
schema: {
$ref: "#/definitions/Resource"
}
}
],
tags: [
"v1"
]
}
}
},
info: {
title: "API",
version: "1.0"
},
produces: [
"application/json"
],
consumes: [
"application/json"
],
tags: [
{
name: "v1"
}
],
definitions: {
Resource: {
required: [
"hatch_rate",
"num_clients",
"run_time"
],
properties: {
run_time: {
type: "string",
description: "Time to run the test",
example: "10s"
},
num_clients: {
type: "integer",
description: "Number of Users",
example: 1,
minimum: 1
},
hatch_rate: {
type: "integer",
description: "Hatch rate for thetest",
example: 1,
minimum: 1
}
},
type: "object"
}
},
responses: {
ParseError: {
description: "When a mask can't be parsed"
},
MaskError: {
description: "When any error occurs on mask"
}
}
}`

I have the following things that need to be checked:
1) Passing custom type of string like UUID  as the test data.
2) Referencing a dataset or a datasource which will have the custom dataset that can be passed to a particular parameter.

**Reply by Stranger6667** [reply]
I am not aware how exactly `flask-restplus` allows the end-user to modify the schema, but assuming that you want UUID for the `ident` parameter of the `/v1/get/{ident}/{ticker}`  endpoint you need to add the following (see the comment inside code):

```json
{
    "swagger": "2.0",
    "basePath": "/",
    "paths": {
        "/v1/get/{ident}/{ticker}": {
            "parameters": [
                {"name": "ident", "in": "path", "required": true, "type": "string", "format": "uuid"},
                {"name": "ticker", "in": "path", "required": true, "type": "integer"},
            ],
            "get": {
                "responses": {"200": {"description": "Success"}},
                "operationId": "get_hello_world",
                "tags": ["v1"],
            },
        },
        "/v1/post": {
            "post": {
                "responses": {"200": {"description": "Success"}},
                "operationId": "post_hello_world",
                "parameters": [
                    {
                        "name": "payload",
                        "required": true,
                        "in": "body",
                        "schema": {"$ref": "#/definitions/Resource"},
                    }
                ],
                "tags": ["v1"],
            }
        },
    },
    "info": {"title": "API", "version": "1.0"},
    "produces": ["application/json"],
    "consumes": ["application/json"],
    "tags": [{"name": "v1"}],
    "definitions": {
        "Resource": {
            "required": ["hatch_rate", "num_clients", "run_time"],
            "properties": {
                "run_time": {
                    "type": "string",
                    "description": "Time to run the test",
                    "example": "10s",
                },
                "num_clients": {
                    "type": "integer",
                    "description": "Number of Users",
                    "example": 1,
                    "minimum": 1,
                },
                "hatch_rate": {
                    "type": "integer",
                    "description": "Hatch rate for thetest",
                    "example": 1,
                    "minimum": 1,
                },
            },
            "type": "object",
        }
    },
    "responses": {
        "ParseError": {"description": "When a mask can't be parsed"},
        "MaskError": {"description": "When any error occurs on mask"},
    },
}
```

AND add the code from my comment above to your Schemathesis setup. How do you run Schemathesis, btw? It has some differences between CLI & Python tests.

Then, it will generate valid UUIDs for that parameter.

> Referencing a dataset or a datasource which will have the custom dataset that can be passed to a particular parameter.

Do you mean providing known values for some parameters?

**Reply by gouseferoz** [reply]
I am using the pytest way of running the schemathesis using the command `pytest -v main.py`

```python
import schemathesis
from hypothesis import settings, Phase, strategies
from schemathesis import register_string_format

register_string_format("ident", strategies.uuids().map(str))
register_string_format("run_time", strategies.uuids().map(str))

# schema = schemathesis.from_uri("http://127.0.0.1:5000/swagger.json")

schema = schemathesis.from_path("schema.json")

@schema.parametrize()

def test_api(case):
    case.call_and_validate()
```
> AND add the code from my comment above to your Schemathesis setup. How do you run Schemathesis, btw? It has some differences between CLI & Python tests.

I have tried it and I see exceptions during test execution

```
    raise ValidationError("Invalid schema") from exc
E   jsonschema.exceptions.ValidationError: Invalid schema
======================================================================================================== short test summary info =========================================================================================================
ERROR main.py - jsonschema.exceptions.ValidationError: Invalid schema
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================================================================================ 1 error in 0.30s ============================================================================================================
```

> Referencing a dataset or a datasource which will have the custom dataset that can be passed to a particular parameter.

What i am thinking is that, considering the above case, if i want to pass/select the inputs for the `ticker` param from a datasource: which can be a txt file with all the possible data something like

```
register_string_format("ident", path('data/fields.txt'))
```

**Reply by Stranger6667** [reply]
Thanks for sharing your setup!

In the `register_string_format` call, you need to provide the `format` name instead of the parameter name. For example:

```python
register_string_format("uuid", strategies.uuids().map(str))
```

And all parameters that have `"format": "uuid"` in their definitions will automatically use the specified strategy. So, `ident` definition might look like this:

```json
{"name": "ident", "in": "path", "required": true, "type": "string", "format": "uuid"}
```

> I have tried it and I see exceptions during test execution

I checked the error message and it turns out that there is a bug (which is fixed by #1344) that doesn't allow Schemathesis to report a more helpful error. The problem is with numeric status codes - the Open API spec requires them to be strings, and Schemathesis is strict by default (it [might change](https://github.com/schemathesis/schemathesis/issues/1304), though). 
To make Schemathesis work, you can either:
- Change those statuses to strings (I assume that `flask-restplus` generates them correctly and it is purely my mistake in the schema above);
- Add `validate_schema=False` keyword argument to the `schemathesis.from_path` call;

I updated my comment above, so the schema is valid JSON.

> What i am thinking is that, considering the above case, if i want to pass/select the inputs for the ticker param from a datasource: which can be a txt file with all the possible data something like

You can totally do it with `sampled_from` and using `register_string_format`:

```python

idents = sorted(...)  # A list of strings, note the `sorted` call

register_string_format("ident-format", strategies.sampled_from(idents))
```

Then the format in the schema:

```json
{"name": "ident", "in": "path", "required": true, "type": "string", "format": "ident-format"}
```

Alternatively, you could use `schema.given` to avoid the `format` keyword altogether:

```python

idents = sorted(...)  # A list of strings, note the `sorted` call

@schema.parametrize()
@schema.given(data=strategies.data())
def test_api(data, case):
    if case.path == "/v1/get/{ident}/{ticker}":
        case.path_parameters["ident"] = data.draw(
            strategies.sampled_from(idents)
        )
    case.call_and_validate()
```

In the example above, we randomly choose an `ident` value inside the test body replacing the generated one. It is slightly less efficient, because the replaced value is not actually needed, but gives a bit more flexibility.

**NOTE**: To use `sampled_from` you need to pass a sorted collection to it.

I hope that it helps :) Let me know if I can provide more info

**Reply by gouseferoz** [reply]
@Stranger6667 Thanks for the information. I can now understand using them. 

I am using
`print(case.as_text_lines())` to see what data is being used for test. Is this the way to display the test inputs or is there any other way?

**Reply by Stranger6667** [reply]
You're welcome :)

> I am using print(case.as_text_lines()) to see what data is being used for test. Is this the way to display the test inputs or is there any other way?

Yes, there is a [closely related thread](https://gitter.im/schemathesis/schemathesis?at=61a686908f98cb08119b4d35) in gitter with a code sample. Generally, #701 would solve the issue by using VCR cassettes, but it is not implemented yet

**Reply by gouseferoz** [reply]
@Stranger6667 
Is it possible to register_string_format on `parameter` name?

**Reply by Stranger6667** [reply]
The `register_string_format` works only for the JSON Schema's (and OpenAPI) `format` keyword.

There is also #8, but it is not implemented yet as I am not sure about ergonomic API for this purpose
